### ISSUE TITLE
Unify all TS types used in different parts of AsyncAPI project into one document

### ISSUE THREAD

**Issue by aeworxet** [asyncapi/asyncapi-react#255]
Working on [#130](https://github.com/asyncapi/parser-js/issues/130) in `parser-js` repo and finding slightly different versions of `d.ts` files, when gathering information on definition of types each element uses, I thought it would be logical to unify all TS types, used throughout different parts of the AsyncAPI project, into one main `types.d.ts` file, to have one single source of truth:
https://github.com/asyncapi/parser-js/blob/master/types.d.ts
https://github.com/asyncapi/asyncapi-react/blob/master/library/src/types.ts

As I understand, TS types used in `parser-js`, are the most up to date with current specification and requirements, so I guess for unification purposes should be used those.

And maybe this main project-wide AsyncAPI's `types.d.ts` file, after that, should even be uploaded to the [**DefinitelyTyped**](https://github.com/DefinitelyTyped/DefinitelyTyped) repository, into its own directory `asyncapi`, to ease the life of every new TS developer who starts adopting AsyncAPI spec.

```
npm install --save-dev @types/asyncapi
```
would look nice.

**Comment by derberg** [comment]
@magicmatatjahu what do you think about this. Topic `DefinitelyTyped` is something for the future maybe. I more interested in "unification". I guess https://github.com/asyncapi/asyncapi-react/blob/master/library/src/types.ts should be removed and types from the parser should be used. What do you think? I mean what is the point of maintaining own types

**Comment by magicmatatjahu** [comment]
@derberg @aeworxet I had similar idea few months ago. Problem with types in component inside is that we have here only inlined types for schema - in parser we have types for functions like `getDescription` etc. Internally in component we still use schema by `schema.info.contact.email` etc., so if we want unification now, then we must update also retrieving schema's data inside component.

https://github.com/asyncapi/generator-model-sdk should helps us to have unificated types for schema, but when we will generate types... I don't know.

**Comment by derberg** [comment]
my assumption was that anyway some refactor in components would be necessary

**Comment by magicmatatjahu** [comment]
@derberg But I don't know when we will have time to unificate types. Maybe when we will start `@asyncapi/ui` repo.

**Comment by derberg** [comment]
@magicmatatjahu open your mind, it is not just me and you here 😄 maybe @aeworxet wants to work on this 😉 

I think we should just focus on setting a direction here, and direction for all of us I think is pretty obvious, duplication sucks and if technically it is already possible to reuse types provided by the Parser, then the future direction is for sure to use it 😄

**Comment by aeworxet** [comment]
TL;DR I will for sure use `RegExp`s to pull out individual types from `types.d.ts` so for now I'll just keep in mind requirement of future reusability while solving [#130](https://github.com/asyncapi/parser-js/issues/130) and will share when I get new ideas.


As [#130](https://github.com/asyncapi/parser-js/issues/130) is a very complex task (for example, only assuring that `serverNames()` returns `string[]` required 10 lines of code, by checking main container with `isArray()` and that each element in it has typeof `string` - and it was only one single type) and I will for sure use `RegExp`s to pull out individual types from `types.d.ts` to be able to compare to them (tests should not go through massive rewriting each time a single type changes, so getting individual type from a declaration file, being able to determine which reusable chunk of code to use, and use it to test it - should be automated as much as possible), so for now I'll just keep in mind requirement of future reusability while solving it and will share new thoughts when I get them.

Anyway, at least I had drawn attention to this question, so I think @jonaslagoni should also read this (I'm basing my work on his `JSDoc -> TS type` conversion and regexing of `types.d.ts` after that) because he might envisage future TS types' unification and reusability (I see parser's `types.d.ts` in `generator-model-sdk`, so maybe part of `./parser-js/types.d.ts` can be used in `./generator-model-sdk/src/models/CommonSchema.ts`, for example) also.

**Comment by derberg** [comment]
@aeworxet does it make sense to work on #130 before this one? if react component would use the majority of types from the parser, then it would be satisfactory to have a test in parser, that users react to render documentation that is checked using snapshot testing. Wouldn't this be sufficient for making sure types are good and do not break anything?

**Comment by aeworxet** [comment]
Oh, so I overcomplicated the task, trying to ensure every type manually?..
Well, every cloud has a silver lining - otherwise, without digging into the internals, I wouldn't raise this question, which is useful for strategy of development. :slightly_smiling_face:

**Issue [https://github.com/asyncapi/parser-js/issues/130](https://github.com/asyncapi/parser-js/issues/130) is postponed until [React component](https://github.com/asyncapi/asyncapi-react) is refactored to use the majority of types from the [parser-js](https://github.com/asyncapi/parser-js).**

**Comment by magicmatatjahu** [comment]
@aeworxet Hi! We published `1.0.0-next.1` version of component and we switched to parser API. Is this issues still relevant?

**Comment by magicmatatjahu** [comment]
@aeworxet Ok, we still have these types in `master` branch and if we will merge next to master we will still have them in git history. I close issue, if you think that we should take some actions here, please reopen it :)
